crates/bifrost/src/types.rs

```diff
@@ -59,18 +75,107 @@ pub struct FindTailAttributes {
     // TODO: consistent_read: bool,
 }
 
+/// Represents the state of the tail of the loglet.
+#[derive(Clone, Debug)]
+pub enum TailState<Offset = Lsn> {
+    /// Loglet is open for appends
+    Open(Offset),
+    /// Loglet is sealed. This offset if the durable tail.
+    Sealed(Offset),
+}
+
+impl<Offset: SequenceNumber> TailState<Offset> {
+    pub fn map<F, T>(self, f: F) -> TailState<T>
+    where
+        F: FnOnce(Offset) -> T,
+    {
+        match self {
+            TailState::Open(offset) => TailState::Open(f(offset)),
+            TailState::Sealed(offset) => TailState::Sealed(f(offset)),
+        }
+    }
+
+    pub fn is_sealed(&self) -> bool {
+        matches!(self, TailState::Sealed(_))
+    }
+
+    pub fn offset(&self) -> Offset {
+        match self {
+            TailState::Open(offset) | TailState::Sealed(offset) => *offset,
+        }
+    }
+}
+
 #[cfg(test)]
 mod tests {
     use crate::loglet::LogletOffset;
     use crate::types::LsnExt;
     use restate_types::logs::{Lsn, SequenceNumber};
 
     #[test]
-    fn convert_invalid_lsn_into_invalid_offset() {
+    fn lsn_to_offset_oldest() {
+        let lsn = Lsn::OLDEST;
+        let offset = lsn.into_offset(Lsn::OLDEST);
+        assert_eq!(offset, LogletOffset::OLDEST);
+
+        // should work if base_lsn is even zero
+        let lsn = Lsn::from(1);
+        let offset = lsn.into_offset(Lsn::from(0));
+        // offset would be 2 -- lsn(0), lsn(1) -->  offset(1), offset(2)
+        assert_eq!(offset, LogletOffset::from(2));
+    }
+
+    #[test]
+    fn lsn_to_offset() {
+        // validate we still return oldest if lsn is smaller than base_lsn.
         let lsn = Lsn::INVALID;
+        let offset = lsn.into_offset(Lsn::from(100));
+        assert_eq!(offset, LogletOffset::OLDEST);
+
+        let lsn = Lsn::OLDEST;
+        let offset = lsn.into_offset(Lsn::from(100));
+        assert_eq!(offset, LogletOffset::OLDEST);
+
+        let lsn = Lsn::from(99);
+        let offset = lsn.into_offset(Lsn::from(100));
+        assert_eq!(offset, LogletOffset::OLDEST);
 
+        // base_lsn -> oldest
+        let lsn = Lsn::from(100);
+        let offset = lsn.into_offset(Lsn::from(100));
+        assert_eq!(offset, LogletOffset::OLDEST);
+
+        let lsn = Lsn::from(101);
+        let offset = lsn.into_offset(Lsn::from(100));
+        assert_eq!(offset, LogletOffset::OLDEST.next());
+
+        // validate we are saturating correctly
+        let lsn = Lsn::MAX;
         let offset = lsn.into_offset(Lsn::OLDEST);
+        assert_eq!(offset, LogletOffset::MAX);
```

Comment:
Why is this a question of saturating operations? Isn't it a question of applying operations (minus offset, plus oldest) in the right order?